assets/src/block-editor/components/with-media-library-notice.js

```diff
@@ -57,47 +56,82 @@ export default ( InitialMediaUpload, minImageDimensions ) => {
 		 */
 		initFeaturedImage() {
 			const FeaturedImageSelectMediaFrame = getSelectMediaFrame( FeaturedImageToolbarSelect );
+
+			const FeaturedImageLibrary = wp.media.controller.FeaturedImage.extend( {
+				defaults: {
+					...wp.media.controller.FeaturedImage.prototype.defaults,
+					date: false,
+					filterable: false,
+					// Note: These suggestions are shown in the media library image browser.
+					suggestedWidth: EXPECTED_WIDTH,
+					suggestedHeight: EXPECTED_HEIGHT,
+				},
+			} );
+
 			this.frame = new FeaturedImageSelectMediaFrame( {
 				allowedTypes: this.props.allowedTypes,
-				button: {
-					text: __( 'Select', 'amp' ),
-					close: false,
-				},
-				states: [
-					new wp.media.controller.Library( {
-						title: __( 'Choose image', 'amp' ),
-						library: wp.media.query( { type: 'image' } ),
-						multiple: false,
-						date: false,
-						priority: 20,
-						// Note: These suggestions are shown in the media library image browser.
-						suggestedWidth: EXPECTED_WIDTH,
-						suggestedHeight: EXPECTED_HEIGHT,
-					} ),
-				],
+				state: 'featured-image',
+				states: [ new FeaturedImageLibrary(), new wp.media.controller.EditImage() ],
 			} );
 
+			this.frame.on( 'toolbar:create:featured-image', function ( toolbar ) {
+				/**
+				 * @this wp.media.view.MediaFrame.Select
+				 */
+				this.createSelectToolbar( toolbar, {
+					text: wp.media.view.l10n.setFeaturedImage,
+					state: this.options.state,
+				});
+			}, this.frame );
+
+			this.frame.on( 'open', this.onOpen );
```

Comment:
When the featured image has been changed  but not saved, it was not being pushed to the front of the image collection. Hooking into the `open` event here allows for that to happen.

Reply:
@pierlon Can you explain why we need the image pushed to the front of the collection when this happens or what behavior this fixes?

Reply:
I guess this is trying to match the core behavior - I see now that it moves the selected item to the top of the list which is nice. It does something funny though if you close the modal with another image selected, then reopen it - the wrong image is moved to the front. Screencast: https://share.getcloudapp.com/12u1n5Yo

Testing with the plugin disabled I _see the exact same behavior_, so I guess this is fine for now (although it would be great to fix this bug upstream - probably if the user cancels the selection, the original item should be reselected)

Reply:
Sorry for the late reply; yes it was to match the behavior of Core.

> It does something funny though if you close the modal with another image selected, then reopen it - the wrong image is moved to the front.

Interesting... yes it would be best to have the bug fixed upstream. I'll file an issue in the Gutenberg repo.